packages/@uppy/companion/src/server/Uploader.js

```
@@ -553,6 +551,13 @@ class Uploader {
         },
       }
 
+      if (this.options.companionOptions.tusDeferredUploadLength && !isFileStream) {
+        tusOptions.uploadLengthDeferred = true
+      } else {
+        tusOptions.uploadLengthDeferred = false
+        tusOptions.uploadSize = this.size
```

If Companion is configured to not use deferred lengths but also don't have a file size, it should raise a meaningful error. I don't see this case handled here and I can't tell what will happen in those cases with the PR's current code. But I would assume some weird behavior occurs that is hard to debug. A meaningful error helps developers and operators to troubleshoot this quickly.

---

Any hints how to test this behavior? (As in which provider to use)

What kind of error are you thinking of? Just throw an exception?

---

> Any hints how to test this behavior? (As in which provider to use)

You can test it with a HTTP resource that doesn't respond with a Content-Length header. https://httpbin.org/#/ provides two endpoints `/stream-bytes/{n}` and `/stream/{n}` (under dynamic data) that don't seem to include this header in my tests.



> What kind of error are you thinking of? Just throw an exception?

That is better answered by @mifi or @Murderlon.

---

One way to test it is to copy `.env.example` into `.env`, run `yarn dev:with-companion`, and use the `@uppy/url` plugin to test Companion locally with something like httpbin.org.

We don't have a strict format for errors so I suppose anything informative would help.

---

When running Companion like this 

```
COMPANION_TUS_DEFERRED_UPLOAD_LENGTH=false COMPANION_ENABLE_URL_ENDPOINT=true COMPANION_CLIENT_ORIGINS=true yarn run dev:with-companion
```

and try to upload https://httpbin.org/stream/20 I get this error:
```
companion: 2025-01-08T19:44:14.919Z [error] edd2bc3b uploader.error Error: tus: cannot automatically derive upload's size from input. Specify it manually using the `uploadSize` option or use the `uploadLengthDeferred` option
    at /home/domme/dev/sources/owncloud/uppy/packages/@uppy/companion/node_modules/tus-js-client/lib.es5/upload.js:668:31
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```
which basically already hints at the `COMPANION_TUS_DEFERRED_UPLOAD_LENGTH` option which needs to be set to `false` explicitly... 

I've added an explicit error message anyhow - is that what you thought of?

---

Did you test your new error message shows up as expected?

btw I don't think you need to prepend the env vars to `yarn run dev:with-companion`, that's what `.env` is for.

---

Yes, of course - just wanted to record which variables I'm setting to which value.

Well, it's logged to the console and apparently it's published over the websocket to the client - but it's not shown to the user... do you want that? Then I need pointers...